app/channel.go

```diff
@@ -2538,36 +2538,44 @@ func (a *App) MarkChannelAsUnreadFromPost(postID string, userID string, collapse
-			if storeErr != nil && !errors.As(storeErr, &nfErr) {
-				return nil, model.NewAppError("MarkChannelAsUnreadFromPost", "app.channel.update_last_viewed_at_post.app_error", nil, storeErr.Error(), http.StatusInternalServerError)
+		} else if !threadMembership.Following && followThread {
+			opts = store.ThreadMembershipOpts{
+				Following:       true,
+				UpdateFollowing: true,
 			}
-			threadData, storeErr2 := a.Srv().Store.Thread().Get(threadId)
-			if storeErr2 != nil && !errors.As(storeErr2, &nfErr) {
-				return nil, model.NewAppError("MarkChannelAsUnreadFromPost", "app.channel.update_last_viewed_at_post.app_error", nil, storeErr.Error(), http.StatusInternalServerError)
+		}
+		threadMembership, storeErr = a.Srv().Store.Thread().MaintainMembership(user.Id, threadId, opts)
+		if storeErr != nil && !errors.As(storeErr, &nfErr) {
+			return nil, model.NewAppError("MarkChannelAsUnreadFromPost", "app.channel.update_last_viewed_at_post.app_error", nil, storeErr.Error(), http.StatusInternalServerError)
+		}
+		threadData, storeErr2 := a.Srv().Store.Thread().Get(threadId)
+		if storeErr2 != nil && !errors.As(storeErr2, &nfErr) {
+			return nil, model.NewAppError("MarkChannelAsUnreadFromPost", "app.channel.update_last_viewed_at_post.app_error", nil, storeErr.Error(), http.StatusInternalServerError)
+		}
+		if threadData != nil && threadMembership != nil && threadMembership.Following {
+			channel, nErr := a.Srv().Store.Channel().Get(post.ChannelId, true)
+			if nErr != nil {
+				return nil, model.NewAppError("MarkChannelAsUnreadFromPost", "app.channel.update_last_viewed_at_post.app_error", nil, nErr.Error(), http.StatusInternalServerError)
 			}
-			if threadData != nil && threadMembership != nil && threadMembership.Following {
-				channel, nErr := a.Srv().Store.Channel().Get(post.ChannelId, true)
-				if nErr != nil {
-					return nil, model.NewAppError("MarkChannelAsUnreadFromPost", "app.channel.update_last_viewed_at_post.app_error", nil, nErr.Error(), http.StatusInternalServerError)
-				}
-				threadMembership.UnreadMentions, err = a.countThreadMentions(user, post, channel.TeamId, post.UpdateAt-1)
-				if err != nil {
-					return nil, err
-				}
-				_, nErr = a.Srv().Store.Thread().UpdateMembership(threadMembership)
-				if nErr != nil {
-					return nil, model.NewAppError("MarkChannelAsUnreadFromPost", "app.channel.update_last_viewed_at_post.app_error", nil, nErr.Error(), http.StatusInternalServerError)
-				}
+			threadMembership.UnreadMentions, err = a.countThreadMentions(user, post, channel.TeamId, post.UpdateAt-1)
+			if err != nil {
+				return nil, err
+			}
+			_, nErr = a.Srv().Store.Thread().UpdateMembership(threadMembership)
+			if nErr != nil {
+				return nil, model.NewAppError("MarkChannelAsUnreadFromPost", "app.channel.update_last_viewed_at_post.app_error", nil, nErr.Error(), http.StatusInternalServerError)
+			}
+			if opts.UpdateFollowing {
```

Comment:
It seemed to me that we didn't update unread mentions when we mark a thread as unread from a post
(we did that only when there was no thread membership).
At least unit tests showed that much. I don't know at which point we fix this when returning the thread to the user,
because it seems that the user receives the correct amount of mentions...
So I'm re-calculating mentions and updating the `ThreadMembership` regardless of follow state,
but send a WS socket event only when the thread is to be followed.

Anyway I just want to draw your attention upon this change.

Reply:
It looks like the only place where this is called from is `setPostUnread` in api4/post.go, and it always sets `followThread` to be false. Which means `else if !threadMembership.Following && followThread` never evaluates to true.

So in that case, we will be calling `MaintainMembership` with default opts rather than not calling it at all, in the original scenario. Is that correct?